Sbn osc

gputnam opened the pull request:
Updates to NumuSelection in SBNOsc directory. Also some stuff in a NuMuSelection directory inside sbncode/, and a directory inside sbnanalysis/ with some code to be ported to SBNOsc.

Currently, the sbnanalysis/NuMuSelection directory is not compiled (and it would fail as it is made with the old core code). I'm planning to port the remaining code and then delete it.

Commit:
add in icarus to deps so building works inside mrb

gputnam committed:
Add in NuMuSelection directory. Currently just has a fhicl to run pandora on uboone MC.

gputnam committed:
Add in NuMuSelection in sbnanalysis to test stuff out. Initial pandora to truth matching module

gputnam committed:
improve bash completion (now usable)

gputnam committed:
temporary workaround for MC only containing cosmics

gputnam committed:
Merge branch 'master' into feature/gputnam_numu_selection

gputnam committed:
remove txt file

gputnam committed:
add in custom cpp class for reconstruciton event info and generate ROOT dictionary

gputnam committed:
add in placeholder CMacro for calculating chi2

gputnam committed:
add in data type storing stuff associated w/ MCTruth

gputnam committed:
add in FV algorithm

gputnam committed:
add in producer of AssocTruthInfo -- right now just does truth FV

gputnam committed:
ignorize sbnana slection code. Add in config

gputnam committed:
macros and plotter scripts for generating oscillation ratio plot and calculating statistical chi2

gputnam committed:
add in util

gputnam committed:
update CMAkelists

gputnam committed:
split util into .cc file

gputnam committed:
clean up recob

gputnam committed:
move macros into pyRoot avilable classes

gputnam committed:
cmake from last commit

gputnam committed:
add classdef for assoc

gputnam committed:
update fhicl

gputnam committed:
add fhicl for fiducial volume and associated truth stuff

gputnam committed:
remove definition from header file to fix compilation issues

gputnam committed:
Merge branch 'master' into feature/gputnam_numu_selection

gputnam committed:
merge sbn osc updates to core

gputnam committed:
remove NuMuSelection dir during refactor into SBNOsc

gputnam committed:
compile with uboone code

gputnam committed:
refactor config and add in cut for fiducial volume

gputnam committed:
modify fiducial volume class for general detector setup

gputnam committed:
add all typedefs

gputnam committed:
update fhicl for generating FV stuff

gputnam committed:
get truth information correctly and add print option

gputnam committed:
add config files with FV information

gputnam committed:
fix typo and bug in volumes

gputnam committed:
fixup compilation so NumuSelection gets necessary uboonecode libraries

gputnam committed:
cleanup config setting

gputnam committed:
make sample config files correct

gputnam committed:
add in ECCQE utility function

gputnam committed:
add in optional reco vertex cut and add in some histograms

gputnam committed:
add in active volume in configuration

gputnam committed:
add in oscillation probability utility

gputnam committed:
helper function to print out volumes

gputnam committed:
enable running print helper

gputnam committed:
change default config

gputnam committed:
add (stub) min length cut on associated lepton

gputnam committed:
add in visible energy value. Implementation may be incorrect -- has a systematic bias relative to  truth

gputnam committed:
add utilities for getting mass from PDG and for matching tracks+showers to Nu vertex

gputnam committed:
switch to have user provide PDG Database class. Document parameters. Add in helper for finding length contained in list of fiducial volumes

gputnam committed:
Start calculating a few selection specific values. Fixup length cut

gputnam committed:
clean out old NuMuSelection code

gputnam committed:
add in ROOT dictionary for data class

gputnam committed:
address corner case in contained length

gputnam committed:
Make const algo objects into static variables in Utilities so users don't need to provide one

gputnam committed:
clean up names

gputnam committed:
fix whitespace

gputnam committed:
add in optional energy smearing to output

gputnam committed:
add in smearing and new cut to NumuConfig

gputnam committed:
fixup interaction value computations

gputnam committed:
fixup cuts

gputnam committed:
make units explicit

gputnam committed:
bug fixes to histos and getting FV information

gputnam committed:
fix config names

gputnam committed:
document default configuration

gputnam committed:
keep track of total length of leptons

gputnam committed:
fix muon pdgid

gputnam committed:
bug fixes in energy calculations

mastbaum commented on sbnanalysis/ana/CMakeLists.txt:
```diff
@@ -18,9 +20,11 @@ link_directories(
 )
 
 add_subdirectory(ExampleAnalysis)
+#add_subdirectory(NuMuSelection)
```
Remove?

gputnam replied:
Fixed

mastbaum commented on sbnanalysis/ana/SBNOsc/NumuSelection.cxx:
```diff
@@ -1,34 +1,142 @@
 #include <iostream>
+#include <math.h> 
```
This is deprecated, prefer `<cmath>`.

gputnam replied:
Fixed

mastbaum commented on sbnanalysis/ana/SBNOsc/NumuSelection.cxx:
```diff
@@ -1,34 +1,142 @@
+  SelectionBase(), 
+  fEventCounter(0), 
+  fNuCount(0), 
+  _interactionInfo(new std::vector<NuMuInteraction>) {}
```
Looks like you're mixing `f` and `_` prefixes, better to use one, and make it `f` for consistency with the framework code). (Also, does this have to be a pointer to a vector?)

gputnam replied:
Switched to "_", which I prefer for class variables and being able to easily distinguish what is a base class variable v. what is an inheritor class variable. Let me know if this is an issue.

mastbaum commented on sbnanalysis/ana/SBNOsc/NumuSelection.cxx:
```diff
@@ -1,34 +1,142 @@
+      double xmin, xmax, ymin, ymax, zmin, zmax;
+      xmin = xmax = ymin = ymax = zmin = zmax = 0;
+      if ((*config)["NumuSelection"].isMember("active_volume")) {
+	xmin = (*config)["NumuSelection"]["active_volume"]["xmin"].asDouble();
```
Missing indentation here and below.

gputnam replied:
Fixed

mastbaum commented on sbnanalysis/ana/SBNOsc/NumuSelection.cxx:
```diff
@@ -37,31 +145,228 @@ bool NumuSelection::ProcessEvent(const gallery::Event& ev, std::vector<Event::In
+    // Get selection-specific info
+    //
+    // Start with the interaction stuff
+    auto intInfo = interactionInfo(ev, mctruth);
```
Use of `auto` makes this bit hard to understand. Also, is it necessary to separate `interactionInfo` which is only used once from `Select` (which has the same inputs)?

gputnam replied:
Removed auto -- I wanted to separate the steps into 2 functions as they perform different tasks and otherwise the body of the "Select" function would be longer than desirable.

mastbaum commented on sbnanalysis/ana/SBNOsc/NumuSelection.h:
```diff
@@ -10,10 +10,21 @@
  */
 
 #include <iostream>
+#include <random>
```
Prefer using `gRandom` (via `TRandom`) everywhere for RNG consistency.

gputnam replied:
Removed random w/ smearing stuff.

mastbaum commented on sbnanalysis/ana/SBNOsc/NumuSelection.h:
```diff
@@ -47,12 +58,103 @@ class NumuSelection : public core::SelectionBase {
   bool ProcessEvent(const gallery::Event& ev, std::vector<Event::Interaction>& reco);
 
+  /** Additional information used by the selection per neutrino interaction */
+  struct NuMuInteraction {
```
I'm not sure about a specialized class for numu interactions, as this will require similarly specialized code downstream to use (e.g. in covariance matrix building). I'd much rather work toward a standardized output structure for any selection, so that samples are interchangeable. This is what `Event::reco` is mean to do, and if `Event::Interaction` isn't the right structure, we should either expand it or use something else there.

gputnam replied:
I kept the specialized class around for the length variables, but I am moving the energy observable into core.

mastbaum commented on sbnanalysis/ana/SBNOsc/NumuSelection.h:
```diff
@@ -47,12 +58,103 @@ class NumuSelection : public core::SelectionBase {
+  /** Configuration parameters */
+  struct Config {
+    art::InputTag truthTag; //!< art tag for MCTruth information
+    art::InputTag mctTag;
```
Please document all members with Doxygen comments (run doxygen to build the docs, and make sure there are no warnings about coverage).

gputnam replied:
Fixed

mastbaum commented on sbnanalysis/ana/SBNOsc/NumuSelection.h:
```diff
@@ -47,12 +58,103 @@ class NumuSelection : public core::SelectionBase {
+    TH2D *h_numu_Vyz; //!< 2D y-z vertex histogram
+  };
+
+  /** cpp distributions for smearing energies */
```
I'm not sure truth-level smearing is needed in the selection... seems like something better done after the fact in analysis using the reconstructed variables. I'd prefer to simplify the code and remove this, unless you feel strongly it's useful here. If kept, this class should get Doxygen-style comments.

gputnam replied:
Removed smearing

mastbaum commented on sbnanalysis/ana/SBNOsc/NumuSelection.h:
```diff
@@ -47,12 +58,103 @@ class NumuSelection : public core::SelectionBase {
+  };
+
+
+  /** Returns whether to apply FV cut on neutrino */
```
Please document input parameters and return values on these methods.

gputnam replied:
Fixed

mastbaum commented on sbnanalysis/ana/SBNOsc/NumuSelection.h:
```diff
@@ -47,12 +58,103 @@ class NumuSelection : public core::SelectionBase {
 
-  /** Configuration parameters */
-  art::InputTag fTruthTag;  //!< art tag for MCTruth information
+  static const unsigned nCuts = 4; //!< number of cuts
```
I'm not sure I understand this bit about the cuts. Why is the length of this vector defined separate from `size()`?

gputnam replied:
I added in the variable to ensure that per-cut values would all be the same length, but I didn't include it for all values. That is fixed now.

mastbaum commented on sbnanalysis/ana/SBNOsc/Utilities.h:
```diff
@@ -26,6 +34,60 @@ void hello();
+ * \param l_energy Lepton energy in GeV
+ * \param energy_distortion Optional energy distortion in GeV
+ * \param angle_distortion Optiona langle distortion 
+ *
```
Could you add a `\returns` comment to these?

gputnam replied:
Fixed

mastbaum commented on sbncode/CMakeLists.txt:
```diff
@@ -5,3 +5,5 @@ add_subdirectory(EventWeight)
 #add_subdirectory(LArSoftConfigurations)
 #add_subdirectory(Utilities)
 
+add_subdirectory(NuMuSelection)
```
There's not much documentation for the LArSoft-side code -- could you please expand it?

gputnam replied:
Added a little -- but also these classes aren't being used in the analysis code. They were intended for just playing around with larsoft stuff. I have removed most of them from the branch.

mastbaum commented on sbncode/NuMuSelection/PandoraTesting_module.cc:
```diff
@@ -0,0 +1,365 @@
+////////////////////////////////////////////////////////////////////////
+
+// Playing around with Pandoraw objects
+// Taken from Rhiannon Jones github
```
I think I found it, but there's no license. Could you inquire about copying/attribution?

gputnam replied:
This wasn't being used in the analysis, so I removed it.

mastbaum commented on sbncode/NuMuSelection/Algorithms/FiducialVolume.h:
```diff
@@ -0,0 +1,88 @@
+ *
+ * @author Marco Del Tutto
+ *
+ * Taken from Marco Del Tutto's UBXSec analysis
```
The licensing situation in LArSoft is unfortunately a mess as far as I can tell. Could you use the code in UBXsec via uboonecode libraries, or implement FV cuts in your own code?

gputnam replied:
Also removed this one, since we only use AABox stuff in the sbnanalysis code.

mastbaum commented on sbnanalysis/ana/CMakeLists.txt:
```diff
@@ -18,9 +20,11 @@ link_directories(
   $ENV{NUSIMDATA_LIB}
   $ENV{LARCOREOBJ_LIB}
   $ENV{LARDATAOBJ_LIB}
+  $ENV{MRB_BUILDDIR}/uboonecode/lib/
```
This'll be different depending on whether `uboonecode` is built in local products or a UPS version is used (and libs are in `$UBOONECODE_LIB`. Would it be possible to detect this, so it works either way?

gputnam replied:
Fixed

mastbaum commented on sbnanalysis/ana/SBNOsc/CMakeLists.txt:
```diff
@@ -16,9 +16,21 @@ add_library(sbnanalysis_SBNOsc_NumuSelection SHARED NumuSelection.cxx)
 add_library(sbnanalysis_SBNOsc_Chi2Sensitivity SHARED Chi2Sensitivity.cxx)
 
+# Add in ROOT dictionary
+ROOT_GENERATE_DICTIONARY(SBNOsc_dict NumuSelection.h LINKDEF linkdef.h)
```
See comments below about `struct NumuInteraction`.

gputnam replied:
Still currently included for length information (and presumably going forward more stuff).

mastbaum commented on sbnanalysis/ana/SBNOsc/NumuSelection.cxx:
```diff
@@ -37,31 +145,228 @@ bool NumuSelection::ProcessEvent(const gallery::Event& ev, std::vector<Event::In
+        _root_histos[select_i].h_numu_Vxz->Fill(nu.Nu().Vx(), nu.Nu().Vz());
+        _root_histos[select_i].h_numu_Vyz->Fill(nu.Nu().Vy(), nu.Nu().Vz());
+      }
+      select_i ++;
```
It seems like it would be more clear to use an index-based loop here.

gputnam replied:
Fixed.

mastbaum commented on sbnanalysis/ana/SBNOsc/NumuSelection.cxx:
```diff
@@ -37,31 +145,228 @@ bool NumuSelection::ProcessEvent(const gallery::Event& ev, std::vector<Event::In
+  // get lepton track
+  int lepton_ind = -1;
+  for (int i = 0; i < mctrack_list.size(); i++) {
+    if (isFromNuVertex(mctruth, mctrack_list[i]) && mctrack_list[i].PdgCode() == 13 && mctrack_list[i].size() > 0) {
```
Also want to check here if it's a primary track?

gputnam replied:
Fixed

mastbaum commented on sbnanalysis/ana/SBNOsc/NumuSelection.cxx:
```diff
@@ -37,31 +145,228 @@ bool NumuSelection::ProcessEvent(const gallery::Event& ev, std::vector<Event::In
+  // first the tracks
+  for (auto const &mct: mctrack_list) {
+    if (isFromNuVertex(mctruth, mct)) {
+      // don't subtract mass for muon
```
A small effect but also true for electrons below, I'd put that in if only for clarity.

gputnam replied:
Fixed.

mastbaum commented on sbnanalysis/ana/SBNOsc/NumuSelection.cxx:
```diff
@@ -37,31 +145,228 @@ bool NumuSelection::ProcessEvent(const gallery::Event& ev, std::vector<Event::In
+  // Also do smearing
+  //
+  // Track energy and masses are in MeV
+  double visible_E = 0.;
```
I think this may need to be refined a bit... it looks like this could include not-so-visible energy from neutral particles, and isn't really a true visible energy since it doesn't account for containment or energy/tracking threshold.

gputnam replied:
Yeah -- I am not positive about the visible energy definition.

I have moved the value into the "Event" class so that it can be used downstream. I have also removed neutral particles from the calculation. However, I don't think the proposal analysis accounted for containment or thresholds.

gputnam replied:
It does account for thresholds actually, so that needs to be added in.

gputnam replied:
Added in a first pass of threshold.

mastbaum commented on sbnanalysis/ana/SBNOsc/NumuSelection.cxx:
```diff
@@ -37,31 +145,228 @@ bool NumuSelection::ProcessEvent(const gallery::Event& ev, std::vector<Event::In
+  }
+
+  // retrun list of cuts
+  return {pass_true_CC, pass_true_CC && pass_FV, pass_true_CC && pass_FV && pass_min_length, pass_true_CC && pass_FV && pass_reco_vertex};
```
It would be nice to clean this up a bit, so that the return value is a bit more generic and makes the association between cuts and pass/fail manifest in a data structure. It doesn't look like the individual combinations actually get used anywhere, just an OR of them.

gputnam replied:
The individual combinations get used for filling up different histograms -- I switched the vector to an array to make the nCuts thing more clear.

What do you mean by more generic?

mastbaum commented on sbnanalysis/ana/SBNOsc/NumuSelection.h:
```diff
@@ -47,12 +58,103 @@ class NumuSelection : public core::SelectionBase {
+    double l_length; //!< total length of lepton track
+    double visible_energy; //!< sum of kinetic energies of particles produced directly in interaction
+    double smeared_visible_energy; //!< visible energy with component particle energies smeared
+    double smeared_eccqe; //!< CCQE energy w/ lepton energy smeared
```
Please include units in doc strings where relevant.

gputnam replied:
Fixed.

mastbaum commented on sbnanalysis/ana/SBNOsc/NumuSelection.cxx:
```diff
@@ -37,31 +145,228 @@ bool NumuSelection::ProcessEvent(const gallery::Event& ev, std::vector<Event::In
 
-  bool selected = !reco.empty();
+  // select true CC events
+  bool pass_true_CC = (nu.CCNC() == simb::kCC && (nu.Mode() == 0 || nu.Mode() == 10) && nu.Nu().PdgCode() == 14);
```
It looks like this eventually is a requirement for the event to pass the selection, which I think is too strict -- the selected sample should include non-CCQE backgrounds like pion mis-IDs or CCQE-like topologies generated through FSI (pion absorption, say), which this will cut.

gputnam replied:
Removed this cut and added in pion track confusion.

mastbaum reviewed (changes requested):
Thanks for putting this together. I've made a first pass at reviewing, mostly focused on coding (as opposed to physics) for now. See comments for details, but some overarching notes:
* Selection: It looks like the selection accepts only true numu CCQE interactions; if so, I think this is too strict.
* Data structure: Output reco-level variables should be in a generic structure so that downstream code doesn't need to specialize for each selection (cf. `NumuInteraction`. Let's discuss what's needed here.
* Documentation: Doxygen-style comments are missing in some places (including in the LArSoft module side). For sbnanalysis, please run doxygen and address any reported issues.
* Style: There are a mix of styles here; would prefer the sbnanalysis/LArSoft/ROOT/Geant4 conventions for e.g. member variable naming.
* Licensing: See comments about borrowed code; will want to sort that out before merging.

gputnam committed:
Cleanup syntax and add documentation

gputnam committed:
Remove energy smearing and visible energy from NuMu class. Add in visible energy stub (currently not set) to core Event class

gputnam commented:
Thanks for the comments -- I will address them and reply to the individual comments as they are addressed. Let me know if there is a better way to do it.

gputnam committed:
remove unused classes and add in warning to remaining stub class

gputnam committed:
add in documentation to private NuMuSelection functions

gputnam committed:
switch over all per-cut values to use arrays

gputnam committed:
fix make to work when uboonecode is not built locally

gputnam committed:
cleanup loop

gputnam committed:
add in promary check

gputnam committed:
units everywhere

gputnam committed:
Reorganize visible energy calculation into core. Moved calculation into util/, and changed top level CMake so that util/ is compiled. Removed compilation of stub .cxx files currently in util/. Changed ProcessorBase to configure more tags -- those required for visible energy calculation.

gputnam committed:
Use base class input tags

gputnam committed:
define visible energy more exactly

gputnam committed:
Add in PDGCharge helper. Don't include neutral particles in visible energy

gputnam committed:
set visible energy in reco object

gputnam committed:
returns -> return in doxygen comments

gputnam committed:
Add in option to turn off truth interaction cut. Remove smearing.

gputnam committed:
Add in confusion of muon/pi+. Trun of truth interaction mode cut by default.

gputnam committed:
update default config

gputnam committed:
Remove truth interaction cut and replace with check on whether (possible pion) primary track exists. Improve checking of whether track in FV.

gputnam committed:
first pass of visible energy threshold

gputnam committed:
Refactor visible energy calculation back into SBNOsc.
To do this, change core/Event so that reco interactions are differnt from truth interactions (i.e. new RecoInteraction class) and add in userdata fields to RecoInteraction. Updated SBNOsc and ExampleSelection selections to new core format. Put energy calculations back into SBNOsc/ and don't build util/ anymore

gputnam committed:
clear up units

gputnam committed:
update default config to account for latest changes

gputnam committed:
fix parsing error

gputnam committed:
fix unit descriptions

gputnam committed:
Merge branch 'sbn-osc' of https://github.com/nuchicago/sbncode into local/energy_refactor

gputnam committed:
fix unitx in config

gputnam committed:
document reco class

gputnam committed:
type typo

gputnam committed:
fix documented units

gputnam committed:
Merge branch 'sbn-osc' into local/energy_refactor

gputnam committed:
add in cut line graph

gputnam committed:
remove visible energy

gputnam committed:
add unfilled stub for CCQE energy

gputnam committed:
add back in util/Interaction class with CCQE energy

gputnam committed:
build interaction class

gputnam committed:
Calculate ccqe energy using function in util/Interaction

gputnam committed:
Merge in RecoInteraction refactor

gputnam committed:
Add in value to RecoInteraction containing index to match to truth objects.

gputnam committed:
fix typo

gputnam committed:
unitx bug

Commit:
re-fix units bug

Commit:
name thigns better

Commit:
update names of histos

Commit:
add in histo

Commit:
add in histo header

Commit:
accomidate multiple active volumes for SBND/Icarus and update config

Commit:
allow negative pions

Commit:
allow negative muons

Commit:
fix config of FV cuts

mastbaum commented:
Merged in 62613fcf6d14ac891baaf00085ce9eb99e7adf1c. Thanks @gputnam!
